com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/adapter/handler/StepRequestHandler.java

```diff
@@ -56,63 +67,109 @@ public StepRequestHandler() {
+                logger.log(Level.SEVERE, failureMessage);
+                return AdapterUtils.createAsyncErrorResponse(response, ErrorCode.STEP_FAILURE, failureMessage);
+            } catch (IndexOutOfBoundsException ex) {
+                final String failureMessage = String.format("Failed to step because the thread %d doesn't contain any stack frame", threadId);
+                logger.log(Level.SEVERE, failureMessage);
+                return AdapterUtils.createAsyncErrorResponse(response, ErrorCode.STEP_FAILURE, failureMessage);
             }
         }
 
         return CompletableFuture.completedFuture(response);
     }
 
     private void handleDebugEvent(DebugEvent debugEvent, IDebugSession debugSession, IDebugAdapterContext context) {
-        StepEvent event = (StepEvent) debugEvent.event;
-        ThreadReference thread = event.thread();
-        debugEvent.shouldResume = false;
-        if (context.getDebugFilters() != null) {
-            if (getPendingStepKind(thread) == Command.STEPIN) {
-                // Check if the step into operation stepped through the filtered code and stopped at an un-filtered location.
-                if (getOriginalStackDepth(thread) + 1 < DebugUtility.getFrameCount(thread)) {
-                    // Create another stepOut request to return back where we started the step into.
-                    DebugUtility.stepOut(thread, debugSession.getEventHub(), context.getDebugFilters().stepFilters);
-                    return;
-                }
-                // If the ending step location is filtered, or same as the original location where the step into operation is originated,
-                // do another step of the same kind.
-                if (shouldFilterLocation(thread, context) || shouldDoExtraStepInto(thread)) {
-                    DebugUtility.stepInto(thread, debugSession.getEventHub(), context.getDebugFilters().stepFilters);
-                    return;
+        Event event = debugEvent.event;
+
+        // When a breakpoint occurs, abort any pending step requests from the same thread.
+        if (event instanceof BreakpointEvent) {
+            long threadId = ((BreakpointEvent) event).thread().uniqueID();
+            StepRequest pendingStepRequest = getPendingStepRequest(threadId);
+            if (pendingStepRequest != null) {
+                DebugUtility.deleteEventRequestSafely(debugSession.getVM().eventRequestManager(), pendingStepRequest);
+            }
+            setPendingStepRequest(threadId, null);
+        } else if (event instanceof ThreadDeathEvent) {
+            long threadId = ((ThreadDeathEvent) event).thread().uniqueID();
+            removeThreadState(threadId);
+        } else if (event instanceof StepEvent) {
+            debugEvent.shouldResume = false;
+            ThreadReference thread = ((StepEvent) event).thread();
+            long threadId = thread.uniqueID();
+            setPendingStepRequest(threadId, null); // clean up the pending status.
```

Comment:
We need to dispose the request before setting it to null, right?

Reply:
No need to dispose it because DebugUtility.stepXXX method also subscribes StepEvent and it will dispose the StepRequest via JDI.